### PR TITLE
Update Apache Commons Text to 1.10.0

### PR DESCRIPTION
**Pull request by ParaskP7**
After an internal discussion (see `peaMlT-2f-p2` and `peaMlT-2f-p2#comment-78`), this PR upgrades Apache's `Commons Text` library to [1.10.0](https://commons.apache.org/proper/commons-text/changes-report.html#a1.10.0) in order fix the possible `StringSubstitutor` related vulnerability in the library.

For more details see: https://commons.apache.org/proper/commons-text/changes-report.html#a1.10.0

And more specifically see release notes entry below:

```
Make default string lookups configurable via system property. Remove
dns, url, and script lookups from defaults. If these lookups are
required for use in StringSubstitutor.createInterpolator(), they must be
enabled via system property. See StringLookupFactory for details.
```

-----

As part of this `commonsText = '1.10.0'` update the below changes were also applied in order to try and strip down as of the Apache Commons dependency as possible:

Remove `Apache Commons` from `Editor` module:

1. [Rename .java to .kt.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/9961c855909b571eb9feef712f64cb86d2cd82d6)
2. [Kotlinize metadata utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/eca153f80d5cf4870fab0dac51de6c2d7127451a)
3. [Update metadata utils kotlin to match its previous java style.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/510d2b0e13992fbd44102bc10d18371fa4954e34)
4. [Resolve detekt warnings for metadata utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/61afa49ac02db5025e035c0fbe32f2da1c230f2b)
5. [Replace unnecessary org apache commons lang dependency for editor.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/fe8246d2a7551e1dacd537ef27837ff020b44c93)

Strip down `Apache Commons` from `WordPress` module:

1. [Remove unnecessary commons lang not implemented exception.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/e950a413313086a4b0b104ec5dfc48d1b62b3352)
2. [Remove unnecessary commons text word utils capitalize.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/ed0c8f427bf5940b7dc38f121740de63cff89388)
3. [Replace commons lang array utils remove with array utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/48c87b341dd2dd3a3a43ac36226d27b0e38e9784)
4. [Remove unnecessary commons lang array utils remove.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/ce6fe7e499573d6e42f09d5166792999792e0c60)
5. [Replace commons lang array utils index of with array utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/762fb23046fdde78b92bd90687a56bf5736d915b)
6. [Remove unnecessary commons lang array utils index of.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/a2fb8535ebfa0e9cebc85a80208b8d749d10dbd2)
7. [Remove unnecessary commons lang date utils add and is same day.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/fcd2cae96c8b55aecde25d2359ac5bd1642b3fcb)

Suppression List:

1. [Suppress detekt warnings for metadata utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/9f4ab790ecbcc741c2b2b66cbf7c6506075ab7ea)
2. [Suppress magic number for stats date utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/066871776af867ed911b9d16fa22d16a6f7e8965)

Refactoring List:

1. [Rename stats related date utils to stats date utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/f1b475ef6d09fdb9ecac17b85ed91ea5948ba85c)
3. [Fix string escape utils import.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/4655c1ffb4854d79563dec1879e287f0b93cedda)
4. [Remove unnecessary remote url local variable.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/b146be1fb904834b7c326e5fd3b03523c4e64caf)

-----

With this change the only Apache's `Commons Lang` and `Commons Text` related utilities that this project is using are the below:

1. `Commons Lang` -> `org.apache.commons.lang3.StringUtils`
    1. `StringUtils.isEmpty(...)` -> See `PostUtils.java` class.
    2. `StringUtils.equals(...)` -> See `PostUtils.java` class.
    3. `StringUtils.join(...)` -> See `MediaBlockType.java` class.
2. `Commons Text` -> `org.apache.commons.text.StringEscapeUtils`
    1. `StringEscapeUtils.unescapeHtml4(...)` -> 22 usages in project.

A quick diff of the previous usage of `commons-lang3-3.8.1` vs. the current usage of `commons-lang3-3.12.0` showed that only the `StringUtils.equals(...)` got updated, see `Step-wise comparison` below:

`commons-lang3-3.8.1`
```
    public static boolean equals(final CharSequence cs1, final CharSequence cs2) {
        ...
        return CharSequenceUtils.regionMatches(cs1, false, 0, cs2, 0, cs1.length());
    }
```

`commons-lang3-3.12.0`
```
    public static boolean equals(final CharSequence cs1, final CharSequence cs2) {
        ...
        // Step-wise comparison
        final int length = cs1.length();
        for (int i = 0; i < length; i++) {
            if (cs1.charAt(i) != cs2.charAt(i)) {
                return false;
            }
        }
        return true;
    }
```

A quick diff of the previous usage of `commons-text-1.1` vs. the current usage of `commons-text-1.10.0` showed that only the inner to `translate(...)` method of `StringEscapeUtils.unescapeHtml4(...)` got slightly updated, see `RuntimeException` vs. `UncheckedIOException` below:

`commons-text-1.1`
```
    public final String translate(final CharSequence input) {
        ...
        } catch (final IOException ioe) {
            // this should never ever happen while writing to a StringWriter
            throw new RuntimeException(ioe);
        }
    }
```

`commons-text-1.10.0`
```
    public final String translate(final CharSequence input) {
        ...
        } catch (final IOException ioe) {
            // this should never ever happen while writing to a StringWriter
            throw new UncheckedIOException(ioe);
        }
    }
```

-----

To test:

1. First, quickly smoke test both, the `WordPress` and `Jetpack` apps, and see if they work as expected.
2. Then, try and be more specific by taking a look at the code changes, associate those to some screens and make sure everything is working as expected, especially on the `Editor` side of things due to the `MetadataUtils` kotlinization.

-----

## Regression Notes

1. Potential unintended areas of impact

The `PostUtils.postHasEdits(...)` method might be producing different results before and after the change, that is, since the `StringUtils.equals(...)` implementation was slightly changed.

9. What I did to test those areas of impact (or what existing automated tests I relied on)

See `To test` section above.

10. What automated tests I added (or what prevented me from doing so)

N/A

-----

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsDateFormatter.kt
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsDateFormatter.kt:103]: Do we need to check if the first char `isLowerCase`? I assume we can just call `titlecase`, but maybe there is an issue with it that I am missing?

**Comment by oguzkocer** [comment]
@ParaskP7 Could you provide some context about why this PR includes Java to Kotlin conversation for some classes? Was that a requirement for this update, or did it maybe make some things easier? I am also unsure about some of the other changes made in this PR and whether they were necessary or not.

I skimmed through the p2 comments and some of the commit messages, and although there is some information there, I am still not sure about the scope of the PR. I reviewed about 70% of the PR, but I wasn't sure how to handle the feedback without this context, so I stopped for now.

**Comment by ParaskP7** [comment]
👋 @oguzkocer !

> @ParaskP7 Could you provide some context about why this PR includes Java to Kotlin conversation for some classes? Was that a requirement for this update, or did it maybe make some things easier? I am also unsure about some of the other changes made in this PR and whether they were necessary or not.

Sure, let me provide some more context below:

1. For the `MetadataUtils` class, which was the only one that got converted to Kotlin, I did that in order to make it possible to get rid the `org.apache.commons.lang3.math.NumberUtils` import and its 3 x `NumberUtils.toInt(width)` usages. By removing this import, I could have then removed the `org.apache.commons:commons-lang3:3.5` dependency from the `libs:editor` module. I did that in order to make the `org.apache.commons:commons-text:1.1` update to `1.10.0` easier, without needing to care about any implications that this might had to the `NumberUtils` on this other `libs:editor` module. You would ask, why did I worry about that, good question, I worried about that because `commons-text` is using `commons-lang3` as a transitive dependency and as such I wanted to make sure that when I update `commons-text` this would have as little footprint on the codebase as possible.
2. I followed the same strategy for the below imports, in the sense that I tried to replace the `Apache Commons` implementation with a `Kotlin` equivalent implementation:
    - `org.apache.commons.lang3.NotImplementedException` -> This was easy.
    - `org.apache.commons.text.WordUtils` -> This was another quick win.
    - `org.apache.commons.lang3.ArrayUtils` -> This required a bit of custom Kotlin code, but nothing too serious (see `ArrayUtils`). I included test to verify correctness.
    - `import org.apache.commons.lang3.time.DateUtils` -> This required a bit of custom Kotlin code, but this code was copy-pasted as it was from the library itself and automatically converted to Kotlin (see `DateUtils`). I didn't include test here because it was a simple copy-paste-convert and the fact that testing `Date` classes can become complicated. But, if you want I can add a test suite for `DateUtils` as well, I was thinking about doing that anyway, just decided in the end that this isn't worth it.

> I skimmed through the p2 comments and some of the commit messages, and although there is some information there, I am still not sure about the scope of the PR. I reviewed about 70% of the PR, but I wasn't sure how to handle the feedback without this context, so I stopped for now.

Imagine that even with my about refactoring, which preceded  the `commons-text` update to `1.10.0`, I was still careful to mention how the `StringUtils.equals(...)` got updated from the previous `1.1` version. I did that to make sure we are aware of that change and make sure that we focus on some additional manual testing around that.

As such, the real goal that I did all that preceding refactor was twofold:
- To attempt and get rid of as such of the `Apache Commons` usage in this project.
- To attempt and make sure that the `commons-text` update will have as less of an impact to this project.

-----

Before, the `Apache Commons` usage in this project was as extended as:
1. `org.apache.commons.lang3.math.NumberUtils`
2. `org.apache.commons.lang3.NotImplementedException`
3. `org.apache.commons.text.WordUtils`
4. `org.apache.commons.lang3.ArrayUtils`
5. `import org.apache.commons.lang3.time.DateUtils`
6. `org.apache.commons.lang3.StringUtils`
7. `org.apache.commons.text.StringEscapeUtils`

With this PR, the `Apache Commons` usage in this project is as little as:
1. `org.apache.commons.lang3.StringUtils` x 2 imports
2. `org.apache.commons.text.StringEscapeUtils` x 22 imports

As some point, it would be great if we could strip-off the `org.apache.commons.lang3.StringUtils` as well and be left only with the `org.apache.commons.text.StringEscapeUtils` import and its `StringEscapeUtils.unescapeHtml4(...)` usages in this project.

-----

Let me know if the above helps providing some more context Oguz and thank you for reviewing this PR. 🙏 🙇

File: WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt
Comment (oguzkocer) [inline WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt]: In my opinion, using [assertFailsWith](https://kotlinlang.org/api/latest/kotlin.test/kotlin.test/assert-fails-with.html) would be more declarative than annotation properties. However, that might be up to personal taste. Either way, may I suggest removing the code that'll not be executed to simplify the test?

```suggestion
        ArrayUtils.remove(stringArray, -1)
```

File: WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt
Comment (oguzkocer) [inline WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt]: Similar to my previous comment, may I suggest removing the code that'll not be executed?

```suggestion
        ArrayUtils.remove(stringArray, 5)
```

**Review by oguzkocer — COMMENTED** [review]
@ParaskP7 Thank you for the extra context and handling this change 🙇‍♂️ 

I've left some minor suggestions, but none of them are blocking. I'll test this tomorrow, take another quick look and approve the PR unless anything comes up during testing.

Besides the line comments, I wanted to suggest excluding the transitive dependencies you're worried about from `org.apache.commons:commons-text` so we can avoid developers accidentally using it. You mentioned `commons-lang3`, but I guess we are not ready to exclude that one yet and I don't know if there are any others, but I thought I'd bring it up.

---

I had some comments about `MetadataUtils.kt`, but decided to skip them for the review. I feel we should at least get rid of the `addMetadataProperty` methods, because they are more harmful than helpful, but I understand that this change is not the point of this PR.

Re: PR strategy; I think there was a middle ground where we could have introduced a Kotlin helper class that we call from the java class instead of including a sub-par Java-to-Kotlin conversion. However I don't think a single instance is a big issue and to keep things practical, I am just leaving this as a footnote rather than a request for a change.

**Comment by ParaskP7** [comment]
👋 @oguzkocer !

> @ParaskP7 Thank you for the extra context and handling this change 🙇‍♂️

Thank YOU, for the review and your very thoughtful comments, they are always very appreciated! 💯 

> I've left some minor suggestions, but none of them are blocking. I'll test this tomorrow, take another quick look and approve the PR unless anything comes up during testing.

Thank YOU! 🙇 

> Besides the line comments, I wanted to suggest excluding the transitive dependencies you're worried about from org.apache.commons:commons-text so we can avoid developers accidentally using it. You mentioned commons-lang3, but I guess we are not ready to exclude that one yet and I don't know if there are any others, but I thought I'd bring it up.

Yea, I would love that too, but unfortunately we can't exclude it for now, because of these `org.apache.commons.lang3.StringUtils` x 2 imports, see `PostUtils.java` and `MedialBlockType.java` files, and it's few usages:
- `PostUtils.java` -> `StringUtils.isEmpty(...)` x 3 -> Should be easy to replace!
- `PostUtils.java` -> `StringUtils.equals(...)` x 8 -> This would require some testing, both automated and manual, before we can feel good about replacing that.
- `MedialBlockType.java` -> `StringUtils.join(...)` x 1 -> Shouldn't be too hard to replace. 🤞 

When the above are all replaced, then we could go forward and exclude this transitive dependency. 💯 

-----

> I had some comments about MetadataUtils.kt, but decided to skip them for the review. I feel we should at least get rid of the addMetadataProperty methods, because they are more harmful than helpful, but I understand that this change is not the point of this PR.

Yea, I am really not sure about the `addMetadataProperty(...)` methods and its usages, I am really not an expect in this area, but if you ask me, all comments are welcome. 🤔

> Re: PR strategy; I think there was a middle ground where we could have introduced a Kotlin helper class that we call from the java class instead of including a sub-par Java-to-Kotlin conversion. However I don't think a single instance is a big issue and to keep things practical, I am just leaving this as a footnote rather than a request for a change.

Yea, you are totally right here, and this is exactly what I did with `ArrayUtils` and `DateUtils`. I should have probably done the same for `MetadataUtils.java` class as well, without converting it to Kotlin altogether. If you feel it might be better if I go and redo this PR, but this time create a helper class instead of converting this class to Kotlin, I'll be more than happy to do it, no problem at all. It shouldn't take me more than an hour to do so, from start to finish with opening a new PR. 💯 

PS: The only reason I went forward with this Java-to-Kotlin conversion for this file was the fact that it seemed a very easy migration, especially since this was a simple utility file. Also, the fact that this utility file was only used by `AztecEditorFragment`, made this decision easier for me, as I knew the extend of its affect. No matter, I can redo it! 👍

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsDateFormatter.kt
Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsDateFormatter.kt:103]: Good question @oguzkocer ! 💯 

Actually, this conversion was done automagically, as a suggestion for the replacement of the deprecated `StringsJVM.capitalize(...)` function. If you click-enter on this deprecated method, the IDE will then suggest you this alternative. As such, this seems the safest choice to me, to trust Kotlin on that:

`String.JVM.kt` -> `L740` -> `@Deprecated`
```
/**
 * Returns a copy of this string having its first letter titlecased using the rules of the specified [locale],
 * or the original string if it's empty or already starts with a title case letter.
 *
 * The title case of a character is usually the same as its upper case with several exceptions.
 * The particular list of characters with the special title case form depends on the underlying platform.
 */
@Deprecated("Use replaceFirstChar instead.", ReplaceWith("replaceFirstChar { if (it.isLowerCase()) it.titlecase(locale) else it.toString() }"))
@DeprecatedSinceKotlin(warningSince = "1.5")
@SinceKotlin("1.4")
@WasExperimental(ExperimentalStdlibApi::class)
@kotlin.internal.LowPriorityInOverloadResolution // To avoid conflicts in function references, as this function was introduced later than common capitalize()
public fun String.capitalize(locale: Locale): String {
    if (isNotEmpty()) {
        val firstChar = this[0]
        if (firstChar.isLowerCase()) {
            return buildString {
                val titleChar = firstChar.titlecaseChar()
                if (titleChar != firstChar.uppercaseChar()) {
                    append(titleChar)
                } else {
                    append(this@capitalize.substring(0, 1).uppercase(locale))
                }
                append(this@capitalize.substring(1))
            }
        }
    }
    return this
}
```

Wdyt? 🤔

**Comment by oguzkocer** [comment]
@ParaskP7 No need to redo it as the comment was just a footnote.

As I mentioned in a recent PR, I don't like to completely skip mentioning something, but I also don't want to hold up PRs, so I leave them like this for future consideration.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsDateFormatter.kt
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsDateFormatter.kt:103]: That is .. _weird_, but I am sure there is a reason. (whether it's a good reason is another question 😓) I guess we leave it as is. 🤷‍♂️

File: WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt
Comment (ParaskP7) [reply WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt]: I love this suggestion and I am all in to apply it here @oguzkocer ! 💯 

However, for us to be able to use [assertFailsWith](https://kotlinlang.org/api/latest/kotlin.test/kotlin.test/assert-fails-with.html) in WPAndroid we would first need to add the `org.jetbrains.kotlin:kotlin-test-junit` dependency via `testImplementation`.

Would you like me to proceed with that? 🤔

PS: WCAndroid is already using this dependency, so we could align WPAndroid like that. Yes, WCAndroid is only using that for its `DateUtilsTest` test suite so far, and one could say that it might be better to not introduce an extra such dependencies for that little of a gain, but I don't see any harm with that, at least I can't think of any, especially since this is a test dependency.

File: WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt
Comment (ParaskP7) [reply WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt]: Thanks for the suggestion @oguzkocer ! 🥇 

Partially done: 442e313c668c3dfd0c073e240ec523f4ccc281d9

File: WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt
Comment (ParaskP7) [reply WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt]: Partially done: 442e313c668c3dfd0c073e240ec523f4ccc281d9

**Comment by ParaskP7** [comment]
👋 @oguzkocer !

> @ParaskP7 No need to redo it as the comment was just a footnote.

Got it, but I also know you're too kind and thus leaving just a footnote! 😊  ❤️ 😄 

> As I mentioned in a recent PR, I don't like to completely skip mentioning something, but I also don't want to hold up PRs, so I leave them like this for future consideration.

Got it, but with me feel free to request changes, or even redo PRs, I won't take it personally and welcome this, especially when I agree on things that could have been done differently, please note this for the future. 🙏 

PS: No matter, this is totally noted for future consideration! 💯

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsDateFormatter.kt
Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsDateFormatter.kt:103]: 😄 I had the same exact thought, but ended up trusting JetBrains and Kotlin on the reason! 😄 

Btw, I can find the reason for you if you like, it is just that I never got down to it, maybe I am trusting JetBrains and Kotlin too much... 😝

Yea, let's leave it as is. 🤷

PS: Another reason I want to leave it as is, it is the fact that next time, when I go and do such replacement I want to have a point of reference that this was already done in the codebase, at least once, to give me that little more of a confidence that this is the correct thing to be doing. For example, see `StringExtensions.kt` and its `String.capitalizeWithLocaleWithoutLint(...)` extension function. In the past, I chose to just deprecate that, maybe next time I'll replace such deprecations instead. As such, when I wrote that code from scratch, I didn't want to use `capitalize(...)` and then suppress this deprecation, I said, let's do it right for new code... 🤞

File: WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt
Comment (oguzkocer) [reply WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt]: @ParaskP7 I don't think my suggestion here is important enough to add a new dependency. However, I am wondering if other developers stay away from these helpers because they need to add a new dependency. So, maybe it's still worth to add it? 🤔 

I will leave the decision to you as I am good with it either way.

File: WordPress/src/main/java/org/wordpress/android/util/DateUtils.kt
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/util/DateUtils.kt]: Note that I left [a comment](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573#discussion_r1029988672) about this in the [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573).

**Review by oguzkocer — APPROVED** [review]
@ParaskP7 I've smoke tested both apps and the changes look good to me. I left one non-blocking comment, but otherwise it's 👍 from me.

File: WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt
Comment (ParaskP7) [reply WordPress/src/test/java/org/wordpress/android/util/ArrayUtilsTest.kt]: Great, I actually agree, thanks for the confirmation on that @oguzkocer ! 💯 

This is now done: 48002a0f8383b9a4fc586252cc6d895e54938371

File: WordPress/src/main/java/org/wordpress/android/util/DateUtils.kt
Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/util/DateUtils.kt]: Agreed @oguzkocer ! 💯 

This is now done: cdf6b8d69e289456db4d7085ca33b5530a12c0e1

**Comment by ParaskP7** [comment]
👋 @oguzkocer !

> @ParaskP7 I've smoke tested both apps and the changes look good to me.

Thank you so much for reviewing and testing the changes, you rock! 🙇 ❤️ 🚀 

> I left one non-blocking comment, but otherwise it's 👍 from me.

I've applied some additional changes based on your comments, let me know what you think of them! 🙏

**Comment by oguzkocer** [comment]
@ParaskP7 Thanks for the latest changes. I think we can merge this after addressing the merge conflict 🙇

**Comment by ParaskP7** [comment]
> @ParaskP7 Thanks for the latest changes. I think we can merge this after addressing the merge conflict 🙇

Awesome, thanks @oguzkocer ! 🙇

**Pull request merged**
